File: solutions/3471-find-the-largest-almost-missing-integer/array.js
// time: O(n)
// space: O(1)

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const largestInteger = (nums, k) => {
    const n = nums.length;

    if (k === 1) {
        const freq = {};
        let result = -1;

        for (const num of nums) freq[num] = (freq[num] || 0) + 1;
        for (const num in freq)
            result = Math.max(result, freq[num] === 1 ? +num : -Infinity);

        return result;
    }

    if (n === k) return Math.max(...nums);

    let last = nums[n - 1];
    let first = nums[0];

    for (let i = n - 2; i >= 0; i--) {
        if (nums[i] === last) {
            last = -1;
            break;
        }
    }

    for (let i = 1; i < n; i++) {
        if (nums[i] === first) {
            first = -1;
            break;
        }
    }

    return Math.max(first, last);
};

File: solutions/3471-find-the-largest-almost-missing-integer/array.cpp
// time: O(n)
// space: O(1)

#include <algorithm>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
  public:
    int largestInteger(vector<int> &nums, int k) {
        const int n = nums.size();
        int result = -1;

        if (k == 1) {
            unordered_map<int, int> freq;
            for (const int num : nums) freq[num]++;
            for (const int num : nums) result = max(result, freq[num] == 1 ? num : -1);
            return result;
        }

        if (k == n) {
            for (const int num : nums) result = max(result, num);
            return result;
        }

        int first = nums[0];
        int last = nums[n - 1];

        for (int i = 1; i < n; i++) {
            if (nums[i] == first) {
                first = -1;
                break;
            }
        }

        for (int i = 0; i < n - 1; i++) {
            if (nums[i] == last) {
                last = -1;
                break;
            }
        }

        return max(first, last);
    }
};
